### PR TITLE
feat(factory): enhance provider factory with suggestions, health chec…

### PR DESCRIPTION
**Pull request by itz-PrathamMittal**
# Pull Request

This commit enhances the AI provider factory with three key features to improve resilience, user experience, and configurability.

## Description

- **Intelligent Suggestion Engine:** If an unknown provider name is requested, the factory now suggests the closest match, guiding users away from simple typos.

- **Health Check & Caching System:** Providers can now implement a 'healthCheck' method. The factory caches healthy provider instances for 5 minutes, preventing repeated calls to unhealthy services and improving performance.

- **Generalized Environment Variable Handling:** The factory now follows a clear priority for model resolution: direct arguments > environment variables > default configuration. This makes the system more flexible and easier to configure across different environments.

- **Testing:** Added a comprehensive test script ('factory-features-advanced-demo.js') to validate all new features and their edge cases. Also added 'tsx' as a dev dependency to enable direct execution of TypeScript scripts.

## Type of Change

- [ ] 🐛 Bug fix (non-breaking change which fixes an issue)
- [x] ✨ New feature (non-breaking change which adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📚 Documentation update
- [ ] 🧹 Code refactoring (no functional changes)
- [ ] ⚡ Performance improvement
- [x] 🧪 Test coverage improvement
- [ ] 🔧 Build/CI configuration change

## Related Issues

- Fixes #
- Related to #

## Changes Made

-
-
-

## AI Provider Impact

- [ ] OpenAI
- [ ] Anthropic
- [ ] Google AI/Vertex
- [ ] AWS Bedrock
- [ ] Azure OpenAI
- [ ] Hugging Face
- [ ] Ollama
- [ ] Mistral
- [ ] All providers
- [ ] No provider-specific changes

## Component Impact

- [ ] CLI
- [ ] SDK
- [ ] MCP Integration
- [ ] Streaming
- [ ] Tool Calling
- [ ] Configuration
- [ ] Documentation
- [ ] Tests

## Testing

- [ ] Unit tests added/updated
- [ ] Integration tests added/updated
- [ ] E2E tests added/updated
- [ ] Manual testing performed
- [ ] All existing tests pass

### Test Environment

- OS:
- Node.js version:
- Package manager:

## Performance Impact

- [ ] No performance impact
- [x] Performance improvement
- [ ] Minor performance impact (acceptable)
- [ ] Significant performance impact (needs discussion)

## Checklist

- [x] My code follows the project's style guidelines
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published

## Summary by CodeRabbit

- New Features
  - Provider health checks with automatic caching and reuse of healthy instances; unhealthy providers are rejected.
  - Smarter handling of unknown provider names with similarity-based suggestions.
  - Model selection now supports environment variable overrides with clear precedence (explicit argument > env var > default), and updated sensible defaults.

- Examples
  - Added an advanced demo script illustrating provider setup, health checks, caching, and model selection.

- Chores
  - Added dependencies for string matching, type definitions, and tooling.

### CODE REVIEW

File: src/lib/factories/providerFactory.ts [hunk 1]
```diff
@@ -75,54 +84,98 @@ export class ProviderFactory {
...
+          isHealthy: true,
+          lastCheck: Date.now(),
+        });
+      }
+
+      this.providerCache.set(cacheKey, providerInstance);
+      return providerInstance;
     } catch (error) {
       logger.error(`Failed to create provider ${providerName}:`, error);
+      this.providerStatus.set(cacheKey, {
```
[the remark this answers is not in the document]
Comment (itz-PrathamMittal) [reply src/lib/factories/providerFactory.ts:193]: Done

File: src/lib/factories/providerFactory.ts [hunk 2]
```diff
@@ -75,54 +84,98 @@ export class ProviderFactory {
...
-        `Unknown provider: ${providerName}. Available providers: ${this.getAvailableProviders().join(", ")}`,
-      );
+      const availableProviders = this.getAvailableProviders();
+      const bestMatch = findBestMatch(
+        normalizedName,
+        availableProviders,
+      ).bestMatch;
+
+      let errorMessage = `Unknown provider: '${providerName}'.`;
+      if (bestMatch.rating > 0.6) {
```
[the remark this answers is not in the document]
Comment (itz-PrathamMittal) [reply src/lib/factories/providerFactory.ts]: Done

File: test/factory-features.test.ts [hunk 3]
```diff
@@ -0,0 +1,159 @@
+// NOTE: This demo imports TypeScript modules (.ts) into a JavaScript file (.js).
+// It must be executed with a loader like tsx.
+// You can run this demo using the pre-configured npm script:
+//
+// pnpm run demo:factory-advanced
```
[the remark this answers is not in the document]
Comment (itz-PrathamMittal) [reply test/factory-features.test.ts]: done

File: src/lib/factories/providerFactory.ts [hunk 4]
```diff
@@ -75,55 +93,108 @@ export class ProviderFactory {
...
 
-      try {
-        // Try as async factory function first (most providers are async functions)
+      const ctor = registration.constructor;
+      // Heuristic: A class has a `prototype` property, a factory function (like an arrow function) does not.
+      // This is more robust than checking constructor names, which can be affected by minification.
+      const isFactoryFunction = !Object.prototype.hasOwnProperty.call(
+        ctor,
+        "prototype",
+      );
```
[the remark this answers is not in the document]
Comment (itz-PrathamMittal) [reply src/lib/factories/providerFactory.ts]: done

**Comment by murdore** [comment]
Closing as part of project audit (2026-03-29). The feature this PR targets was implemented via a different approach in a later release. See docs/project-audit-2026-03-29.md for details.
